proxy/src/transport/tls/cert_resolver.rs

```diff
@@ -0,0 +1,131 @@
+use std::{
+    sync::Arc,
+    time::SystemTime,
+};
+
+use super::{
+    config,
+
+    ring::{self, rand, signature},
+    rustls,
+    untrusted,
+    webpki,
+};
+
+pub struct CertResolver {
+    certified_key: rustls::sign::CertifiedKey,
+}
+
+struct SigningKey {
+    signer: Signer,
+}
+
+#[derive(Clone)]
+struct Signer {
+    private_key: Arc<signature::KeyPair>,
+}
+
+impl CertResolver {
+    /// Returns a new `CertResolver` that has a certificate (chain) verified to
+    /// have been issued by one of the given trust anchors.
+    ///
+    /// TODO: Verify that the public key of the certificate matches the private
+    /// key.
+    pub fn new(
+        trust_anchors: &webpki::TLSServerTrustAnchors,
+        cert_chain: Vec<rustls::Certificate>,
+        private_key: untrusted::Input)
+        -> Result<Self, config::Error>
+    {
+        let now = webpki::Time::try_from(SystemTime::now())
+            .map_err(|ring::error::Unspecified| config::Error::TimeConversionFailed)?;
+
+        // Verify that we were given a valid TLS certificate that was issued by
+        // our CA.
+        parse_end_entity_cert(&cert_chain)
+            .and_then(|cert| {
+                cert.verify_is_valid_tls_server_cert(
+                    &[SIGNATURE_ALG_WEBPKI],
+                    &trust_anchors,
+                    &[], // No intermediate certificates
+                    now)
+                    .map_err(config::Error::EndEntityCertIsNotValid)
+            })?;
+
+        let private_key = signature::key_pair_from_pkcs8(SIGNATURE_ALG_RING_SIGNING, private_key)
+            .map_err(|ring::error::Unspecified| config::Error::InvalidPrivateKey)?;
+
+        let signer = Signer { private_key: Arc::new(private_key) };
+        let signing_key = SigningKey { signer };
+        let certified_key = rustls::sign::CertifiedKey::new(
+            cert_chain, Arc::new(Box::new(signing_key)));
+        Ok(Self { certified_key })
+    }
+}
+
+fn parse_end_entity_cert<'a>(cert_chain: &'a[rustls::Certificate])
+    -> Result<webpki::EndEntityCert<'a>, config::Error>
+{
+    cert_chain.first()
+        .ok_or(config::Error::EmptyEndEntityCert)
+        .and_then(|cert| {
+            webpki::EndEntityCert::from(untrusted::Input::from(cert.as_ref()))
+                .map_err(config::Error::EndEntityCertIsNotValid)
+        })
+}
+
+impl rustls::ResolvesServerCert for CertResolver {
+    fn resolve(&self, server_name: Option<webpki::DNSNameRef>,
+               sigschemes: &[rustls::SignatureScheme]) -> Option<rustls::sign::CertifiedKey> {
+        let server_name = server_name?; // Require SNI
+
+        if !sigschemes.contains(&SIGNATURE_ALG_RUSTLS_SCHEME) {
+            return None;
+        }
+
+        // Verify that our certificate is valid for the given SNI name.
+        parse_end_entity_cert(&self.certified_key.cert).ok()
+            .and_then(|cert| cert.verify_is_valid_for_dns_name(server_name).ok())?;
```

Comment:
Calling `ok()` on these `Result`s discards error information, should we be logging those errors prior to discarding them?

Reply:
Yes, we should add some logging here. In an upcoming change I'll make this so that it is essentially dead code; i.e. Rustls will never ask us to produce a certificate for which we don't have a valid certificate, after I add the protocol detection logic.

Reply:
> Calling ok() on these Results discards error information, should we be logging those errors prior to discarding them?

I added the logging. Thanks for the suggestion.

Reply:
Great!